x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java

```diff
@@ -64,6 +65,8 @@ public FieldAttribute(
         this(source, parent, name, field.getDataType(), field, nullability, id, synthetic);
     }
 
+    // TODO: Should this become private? An explicitly set type will be discarded the moment this is `clone`d, so using this constructor is
```

Comment:
Agree. You could make this package protected (not `protected`), create a `FieldAttributeTestUtils` class in the `test` module but same package as `FieldAttribute` where a `static` `final` method can call this constructor. Also, adding a comment saying it's for testing purposes only.

Reply:
I'll also make `withDataType` throw an error, as that deceptively does nothing.